### DISCUSSION TITLE
transform() option

### DISCUSSION THREAD

**Discussion by rentalhost** [vercel/swr#2405] [Ideas]
I have a use case where I need to _transform_ the data received by the _fetcher_ before it is delivered by the SWR. Well, I could do it in the _fetcher_ itself, but since my fetcher is so simple, I can reuse it in many places, whereas a `transform()` is exceptional.

In my case, the _data_ that I get from the server is disorganized, and I want to keep it in alphabetical order. So I needed to use SWR and do an intermediary `useMemo()` for that, however, I have two copies of pretty much the same information.

Currently:

```ts
const { data: dataOriginal } = useSWR(url, fetcher);
const data = useMemo(() => reorder(dataOriginal), [dataOriginal]);
```

Possible solution modifying fetcher: but in this case, I need to maintain several fetchers or generate greater complexity in them for something that I think would be essential to be the responsibility of the SWR.

```ts
const { data } = useSWR(url, customFetcher);
```

Feature request:

```ts
const { data } = useSWR(url, fetcher, {
    transform: (data) => reorder(data)
});
```

**Comment by wesrog** [comment]
Have you looked into middleware yet? https://swr.vercel.app/docs/middleware

From the docs:
> Middleware receive the SWR hook and can execute logic before and after running it.

Perhaps you can do your transformation in the "before" state as it mentions?

**Reply by rentalhost** [reply]
Hm... Middleware might work, but I believe it makes the process much more complex than what I did (`useMemo`) and what would be ideal (`transform`). Another possibility would be to create a "_fetcher factory_", which might be my next attempt. And perform the transformation within it before send to SWR. Despite everything, it seems to me a compatible idea that SWR has some data transformation capability.
